### PR TITLE
feat(jellyfish-transaction): Add updatemasternode dftx

### LINKED ISSUE #1842: Support `CustomTxType::UpdateMasternode` on Jellyfish
#### What would you like to be added:

> Support https://github.com/DeFiCh/ain/pull/1437

Add `CustomTxType::UpdateMasternode` DfTx support. This includes: 
- RPC typing @marktanrj #1847
- DfTx parsing @marktanrj #1863
- `whale-api` Masternode Index changes @imkven #1869
- DfTx transaction builder @marktanrj #1894

### PR DESCRIPTION
**Pull request by marktanrj**
#### What this PR does / why we need it:

- updatemasternode DfTx parsing

#### Which issue(s) does this PR fixes?:

Fixes part of #1842

### CODE REVIEW

**Review by infinia-yzl — CHANGES_REQUESTED** [review]
Similar to my comment in #1847 , there are a number of important tests that are needed from
- https://github.com/DeFiCh/ain/blob/dfe240ceb8247d36202e21e9ae5034d18a06eeba/test/functional/rpc_updatemasternode.py

**Comment by infinia-yzl** [comment]
@marktanrj please change PR description to "fixes part of", otherwise it will close the entire issue.

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+    // })
+
+    // await container.generate(65)
+
+    // const newAddress2 = await container.getNewAddress()
+    // const secondUpdateTx = await jsonRpc.masternode.updateMasternode(masternodeId, {
+    //   ownerAddress: newAddress2
+    // })
+
+    // await container.generate(65)
```
Comment (cwkang1998) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]: Let's remove unwanted comments

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+      // newAddress,
+      // newAddress2,
+      collateralTxId,
+    })
+
+    // const txn: TransactionSegWit = await builder.masternode.update(updateMasternode, script)
+    const txn: TransactionSegWit = await builder.masternode.update(updateMasternode, script, [additionalVinData])
+    const outs = await sendTransaction(container, txn)
+
+    console.log('debug')
```
Comment (cwkang1998) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]:
```suggestion
```

File: packages/jellyfish-transaction-builder/src/txn/txn_builder.ts [hunk 3]
```diff
@@ -134,6 +134,46 @@ export abstract class P2WPKHTxnBuilder {
...
+        stack: [OP_CODES.OP_RETURN, opDeFiTx]
+      },
+      tokenId: 0x00
+    }
+
+    // const change: Vout = {
+    //   value: total,
+    //   script: changeScript,
+    //   tokenId: 0x00
+    // }
```
Comment (cwkang1998) [inline packages/jellyfish-transaction-builder/src/txn/txn_builder.ts]: Let's remove unused comments in this file as well

File: packages/jellyfish-transaction/__tests__/script/dftx/dftx_masternode/UpdateMasterNode.test.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+  const data = '2319bc7b76b6916d0b61ee7a408d47d33380be746d11a5702bd87bbee5f8d3bd020104141db4d4d5e7d01091545e5573494b2fd1e91215d902041476e76cdb4de312861ce893f307e7e2bd40a8ac2d'
+  // UpdateMasternode.nodeId [LE] (bdd3f8e5be7bd82b70a5116d74be8033d3478d407aee610b6d91b6767bbc1923)
+  // UpdateMasternode.updates [array length] (0x02)
+  // UpdateMasternodeData.updateType [OwnerAddress] (0x01)
+  // UpdateMasternodeAddress.addressType [p2wpkh] (0x04)
+  // UpdateMasternodeAddress.addressPubKeyHash (length 20 = 0x14) (1db4d4d5e7d01091545e5573494b2fd1e91215d9)
+  // UpdateMasternodeData.updateType [OperatorAddress] (0x02)
+  // UpdateMasternodeAddress.addressType [p2wpkh] (0x04)
+  // UpdateMasternodeAddress.addressPubKeyHash (length 20 = 0x14) (76e76cdb4de312861ce893f307e7e2bd40a8ac2d)
```
Comment (cwkang1998) [inline packages/jellyfish-transaction/__tests__/script/dftx/dftx_masternode/UpdateMasterNode.test.ts:50]: Same as previous, let's remove the unused comments.

**Review by cwkang1998 — APPROVED** [review]

**Review by cwkang1998 — CHANGES_REQUESTED** [review]
Unintended approval.

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts (hunk 1 above)
Comment (marktanrj) [reply packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]: Sorry, code is still work in progress. Was blocked and trying to resolve some issues 😅 
Updated PR to draft

File: packages/jellyfish-transaction/__tests__/script/dftx/dftx_masternode/UpdateMasterNode.test.ts (hunk 4 above)
Comment (marktanrj) [reply packages/jellyfish-transaction/__tests__/script/dftx/dftx_masternode/UpdateMasterNode.test.ts:50]: This are comments to describe the dftx transaction similar to how other dftx files does it, should I still remove it?

Comment (cwkang1998) [reply packages/jellyfish-transaction/__tests__/script/dftx/dftx_masternode/UpdateMasterNode.test.ts:50]: Ah sorry, I think I misunderstood this then, let's keep it!

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+    const encoded: string = OP_CODES.OP_DEFI_TX_UPDATE_MASTER_NODE(updateMasternode).asBuffer().toString('hex')
+    const expectedRedeemScript = `6a${encoded}`
+
+    expect(outs.length).toStrictEqual(3)
+    expect(outs[0].value).toStrictEqual(0)
+    expect(outs[0].n).toStrictEqual(0)
+    expect(outs[0].tokenId).toStrictEqual(0)
+    expect(outs[0].scriptPubKey.asm.startsWith('OP_RETURN 446654786d')).toBeTruthy()
+    expect(outs[0].scriptPubKey.hex).toStrictEqual(expectedRedeemScript)
+    expect(outs[0].scriptPubKey.type).toStrictEqual('nulldata')
```
Comment (infinia-yzl) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]: Please assert the entire object / result instead of individual statements.

**Review by infinia-yzl — CHANGES_REQUESTED** [review]

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts (hunk 5 above)
Comment (marktanrj) [reply packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]: I did thought of doing that at first, but there are values that use other methods like `startsWith`, `toBeTruthy`, `toBeGreaterThan`.

If I do that, I'll put them separately like this?

image

Comment (imkven) [reply packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]:
```js
    expect(outs[0]).toStrictEqual(expect.objectContaining({
      value: 0,
      n: 0,
      tokenId: 0,
      scriptPubKey: {
        asm: expect.stringContaining('OP_RETURN 446654786d'),
        hex: expectedRedeemScript,
        type: 'nulldata'
      }
    }))
```
If really want and less strict. I think can use `stringContaining`.

Comment (marktanrj) [reply packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]: Thanks, amended

**Review by imkven — DISMISSED** [review]
LGTM

**Review by infinia-yzl — CHANGES_REQUESTED** [review]
Please add the tests mentioned in the call we had earlier. Other than that, LGTM!

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts [hunk 6]
```diff
@@ -757,4 +758,208 @@ describe('UpdateMasternode', () => {
...
+      await expect(promise).rejects.toThrow(`DeFiDRpcError: 'bad-txns-customtx, UpdateMasternodeTx: Masternode ${masternodeId} is not in 'ENABLED' state (code 16)', code: -26`)
+    }
+
+    {
+      const masternodes = await jsonRpc.masternode.listMasternodes()
+      expect(masternodes[masternodeId]).toStrictEqual(expect.objectContaining({
+        state: 'PRE_ENABLED',
+        ownerAuthAddress: collateralAddress
+      }))
+    }
```
Comment (infinia-yzl) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]: nit: Personally, I would check the state before sending the transaction (and maybe after sending too although I think after is redundant).

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts [hunk 7]
```diff
@@ -757,4 +758,208 @@ describe('UpdateMasternode', () => {
...
+
+    await container.generate(20)
+
+    {
+      const masternodes = await jsonRpc.masternode.listMasternodes()
+      expect(masternodes[masternodeId]).toStrictEqual(expect.objectContaining({
+        state: 'TRANSFERRING',
+        ownerAuthAddress: collateralAddress
+      }))
+    }
```
Comment (infinia-yzl) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]: nit: Personally, I would check the state before sending the transaction (and maybe after sending too although I think after is redundant).

File: packages/jellyfish-transaction-builder/src/txn/txn_builder_masternode.ts [hunk 8]
```diff
@@ -41,4 +45,24 @@ export class TxnBuilderMasternode extends P2WPKHTxnBuilder {
       changeScript
     )
   }
+
+  /**
+   * Build update masternode transaction
+   *
+   * @param {UpdateMasternode} updateMasternode transaction to create
+   * @param {Script} changeScript to send unspent to after deducting the (converted + fees)
+   * @return {Promise<TransactionSegWit>}
```
Comment (infinia-yzl) [inline packages/jellyfish-transaction-builder/src/txn/txn_builder_masternode.ts]: Add `customVinVout` to JSDoc.

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts [hunk 9]
```diff
@@ -757,4 +758,208 @@ describe('UpdateMasternode', () => {
...
+    const updateMasternode: UpdateMasternode = {
+      nodeId: masternodeId,
+      updates: [
+        {
+          updateType: 0x02,
+          address: { addressType: 0x04, addressPubKeyHash: operatorAddressDestKeyHash }
+        }
+      ]
+    }
+    const txn: TransactionSegWit = await builderNode1.masternode.update(updateMasternode, operatorScript)
```
Comment (infinia-yzl) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]: nit: Might improve readability if the 0s (and 1s) are labeled throughout.

```suggestion
    const masternodeOwnerAddress0 = await node[0].rpc.wallet.getNewAddress()
    const masternodeId0 = await node[0].rpc.masternode.createMasternode(masternodeOwnerAddress0)
    await node[0].generate(100) // create masternode and wait for it to be enabled

    const operatorAddress0 = await node[0].rpc.wallet.getNewAddress()
    const operatorAddressDest0: P2WPKH = P2WPKH.fromAddress(RegTest, operatorAddress, P2WPKH)
    const operatorAddressDestKeyHash0 = operatorAddressDest.pubKeyHash
    const operatorScript0 = await fromAddress(operatorAddress, 'regtest')?.script as Script
    await node[0].generate(4)
    await tGroup.waitForSync() // container2 should know about the new masternode

    const updateMasternode0: UpdateMasternode = {
      nodeId: masternodeId0,
      updates: [
        {
          updateType: 0x02,
          address: { addressType: 0x04, addressPubKeyHash: operatorAddressDestKeyHash0 }
        }
      ]
    }
    const txn: TransactionSegWit = await builderNode1.masternode.update(updateMasternode0, operatorScript0)
```

**Review by infinia-yzl — DISMISSED** [review]
Mostly nitpicks. Other than that, LGTM.

File: packages/jellyfish-transaction-builder/src/txn/txn_builder.ts [hunk 10]
```diff
@@ -98,15 +100,21 @@ export abstract class P2WPKHTxnBuilder {
...
    * @param {Script} changeScript to send unspent to after deducting the fees
    * @param {BigNumber} [outValue=0] for the opDeFiTx, usually always be 0.
+   * @param {Array<{ vin: Vin, vout: Vout, prevout: Prevout }>} [customVinVout = []] for custom vin and vout
    */
   async createDeFiTx (
     opDeFiTx: OP_DEFI_TX,
     changeScript: Script,
-    outValue: BigNumber = new BigNumber('0')
+    outValue: BigNumber = new BigNumber('0'),
+    customVinVout: Array<{ vin: Vin, vout: Vout, prevout: Prevout }> = []
```
Comment (fuxingloh) [inline packages/jellyfish-transaction-builder/src/txn/txn_builder.ts]: Can we move this feature out to another PR? Since it can be separated from `updateMasternode` to reduce the impact and properly contextualize the changes.

**Review by imkven — COMMENTED** [review]
LGTM. There have other comments from other stakeholders.

File: packages/jellyfish-transaction-builder/src/txn/txn_builder.ts (hunk 10 above)
Comment (marktanrj) [reply packages/jellyfish-transaction-builder/src/txn/txn_builder.ts]: Ok, I created #1892

**Review by imkven — DISMISSED** [review]

**Comment by marktanrj** [comment]
Give me a moment, trying to see if I can make the vout change/utxo amount more deterministic based on feedback from @delphk

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts (hunk 6 above)
Comment (marktanrj) [reply packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]: Thanks, amended

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts (hunk 7 above)
Comment (marktanrj) [reply packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]: Amended

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts (hunk 9 above)
Comment (marktanrj) [reply packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]: Amended

File: packages/jellyfish-transaction-builder/src/txn/txn_builder_masternode.ts (hunk 8 above)
Comment (marktanrj) [reply packages/jellyfish-transaction-builder/src/txn/txn_builder_masternode.ts]: Thanks for pointing that out, amended

**Comment by marktanrj** [comment]
Ready for review again 😅

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts [hunk 11]
```diff
@@ -0,0 +1,979 @@
...
+        addresses: [await providers.getAddress()],
+        asm: expect.any(String),
+        hex: expect.any(String),
+        reqSigs: 1,
+        type: 'witness_v0_keyhash'
+      }
+    })
+    expect(outs[1].value).toBeGreaterThan(6.99)
+    expect(outs[1].value).toBeLessThan(7)
+  })
```
Comment (delphk) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]: I think that there should be an assertion to check that the reward address has updated correctly (for eg. by checking `listMasternodes`), and similarly for other tests

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts [hunk 12]
```diff
@@ -0,0 +1,979 @@
...
+      ATTRIBUTES: {
+        'v0/params/feature/mn-setowneraddress': 'true',
+        'v0/params/feature/mn-setoperatoraddress': 'true',
+        'v0/params/feature/mn-setrewardaddress': 'true'
+      }
+    })
+    await container.generate(1)
+
+    const balance = await container.call('getbalance')
+    expect(balance >= 2).toBeTruthy()
```
Comment (delphk) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]: any reason why we are asserting the balance to be `>=2`?

Comment (marktanrj) [reply packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]: Oh my bad, I don't need this.
Will remove it 😅

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts (hunk 11 above)
Comment (marktanrj) [reply packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_masternode_update.test.ts]: ok will take note of this in the new PR #1894

**Comment by marktanrj** [comment]
Shifted transaction builder code to #1894

**Review by fuxingloh — APPROVED** [review]
LGTM, this part is clear
